### DISCUSSION TITLE
Issues with my py2 backport of utils.named_pipe: rewrite named pipes (closed, Python 2 is EOL)

### DISCUSSION THREAD

**Discussion by Billy2011** [streamlink/streamlink#3787] [Q&A] [answered]
Travis build on python 2.7 shows many errors, e.g.:
> _________________ ERROR collecting tests/test_api_validate.py __________________
> tests/test_api_validate.py:8: in <module>
>     from streamlink.plugin.api.validate import (
> src/streamlink/plugin/api/validate.py:410: in <module>
>     @validate.register(xml_element)
> ../../../virtualenv/python2.7.15/lib/python2.7/site-packages/singledispatch/__init__.py:218: in register
>     "on an annotated function.".format(**locals())
> E   TypeError: Invalid first argument to `register()`: <class streamlink.plugin.api.validate.xml_element at 0x7fec09592ae0>. Use either `@register(some_class)` or plain `@register` on an annotated function.

Pytest with **test_utils_named_pipe.test_named_pipe** hangs on py2, even if I insert `reader.daemon = True` before the line 
`reader.start()`:

```
    def test_named_pipe(self):
        pipe = NamedPipeWindows()
        reader = ReadNamedPipeThreadWindows(pipe)
        reader.daemon = True
        reader.start()
        pipe.open()
        self.assertEqual(pipe.write(b"foo"), 3)
        self.assertEqual(pipe.write(b"bar"), 3)
        self.assertEqual(pipe.write(b"\0"), 1)
        reader.done.wait(4000)
        self.assertEqual(reader.error, None)
        self.assertEqual(reader.data, b"foobar")
        self.assertFalse(reader.is_alive())
        pipe.close()

```

[https://github.com/Billy2011/streamlink-27/pull/18](https://github.com/Billy2011/streamlink-27/pull/18)

**Comment by bastimeyer** [comment, marked as answer]
You have two separate issues:
https://travis-ci.com/github/Billy2011/streamlink-27/jobs/512750729

1. Failing tests on the `streamlink.plugin.api.validation` module.
   Looks like you didn't update the implicit class inheritance on py2 while merging the changes and singledispatch doesn't know how to handle these types:
   https://github.com/Billy2011/streamlink-27/blame/1e8f7ec8e5505ed716537d4aa229f7241f2dfef7/src/streamlink/plugin/api/validate.py#L100-L112

2. I don't see `thread.daemon = True` in your changes. Without the reader thread being a daemon thread, the entire process will not terminate if the reader thread doesn't halt.

   Looks like it doesn't halt on py2 because you forgot to change the windll.kernel32 API call here and Windows is for some reason returning success, so the reader thread keeps running:
   https://github.com/Billy2011/streamlink-27/blob/4c33a1544f1f7477ac7d0d362769d83ac9c4cfa8/tests/test_utils_named_pipe.py#L51

**Reply by Billy2011** [reply]
> Looks like you didn't update the implicit class inheritance on py2

The py3 code does not run on py2 because the arguments in the order on py2 are not correct. How would you do that?

```
PY3
class union_get:
    def __init__(self, *keys, seq=tuple):
        self.keys = keys
        self.seq = seq

PY2 backport
class union_get:
    def __init__(self, *keys, **kw):
        self.keys = keys
        self.seq = kw.get("seq", tuple)

```

> I don't see thread.daemon = True in your changes.

Sorry, I tried a few things, unfortunately without success, and then I canceled the code change. 

> Looks like it doesn't halt on py2 because you forgot to change the windll.kernel32 API call here

Many thanks, how could I miss that, the 2nd issue is fixed :)

**Reply by bastimeyer** [reply]
What I meant was this, which will make the tests pass:

```diff
diff --git a/src/streamlink/plugin/api/validate.py b/src/streamlink/plugin/api/validate.py
index f28ba3ca..e75ae226 100644
--- a/src/streamlink/plugin/api/validate.py
+++ b/src/streamlink/plugin/api/validate.py
@@ -97,13 +97,13 @@ class attr(SchemaContainer):
     """Validates an object's attributes."""
 
 
-class union_get:
+class union_get(object):
     def __init__(self, *keys, **kw):
         self.keys = keys
         self.seq = kw.get("seq", tuple)
 
 
-class xml_element:
+class xml_element(object):
     """A XML element."""
 
     def __init__(self, tag=None, text=None, attrib=None):
```

```
$ pytest -ra tests/test_api_validate.py 
============================================================= test session starts ==============================================================
platform linux2 -- Python 2.7.18, pytest-4.6.11, py-1.10.0, pluggy-0.13.1
rootdir: /home/basti/repos/streamlink-27
plugins: cov-2.12.1, requests-mock-1.9.3
collected 27 items                                                                                                                             

tests/test_api_validate.py ...........................                                                                                   [100%]

========================================================== 27 passed in 0.04 seconds ===========================================================
```

**Reply by Billy2011** [reply]
Sorry that i misunderstood that.

**Reply by Billy2011** [reply]
All I have now is this one bug that only occurs with Travis: 
```
File "/home/travis/build/Billy2011/streamlink-27/src/streamlink/compat.py", line 59, in <module>
    from backports.functools_lru_cache import lru_cache
ImportError: No module named functools_lru_cache
```
